### PR TITLE
feat(internal): FieldsMap delegation system with FieldsMapCopy() API

### PR DESCRIPTION
**Pull request by amery**
### **User description**
# Performance Optimization: FieldsMap Delegation and Copy Methods

This PR introduces a comprehensive FieldsMap delegation system that significantly optimises field access by intelligently caching and delegating to ancestor loglets, reducing memory usage and improving performance for loglet chains.

## Summary

- **Performance**: Implement intelligent delegation system for FieldsMap operations.
- **Memory**: Add field map sharing between loglets with identical field sets.
- **API**: Add FieldsMapCopy() method for modifiable field maps.
- **Compatibility**: Maintain existing FieldsMap() API contract.
- **Testing**: Comprehensive test coverage with benchmarks.

### 9-Commit Architecture Evolution

**Foundation & Core API:**

- Add sync.Once cached FieldsMap() method to Loglet
- Add FieldsMapCopy with comprehensive delegation optimization (+971/-39 lines)

**Handler Performance Series:**

- Individual optimization commits for each handler (cblog, mock, logrus, logr, zerolog, zap)
- Convert from manual field iteration to cached FieldsMap() access
- Total handler changes: +13/-55 lines across 6 files

**Advanced Features:**

- Implement FieldsMap parent delegation optimization

### Core FieldsMap Architecture (internal/loglet.go)

**Thread-Safe Caching System (+150/-11 lines):**

- Add `sync.Once` field for one-time map initialization
- Implement `FieldsMap()` with intelligent delegation
- Add `FieldsMapCopy()` method for modifiable field maps with excess capacity
- Enhanced `Copy()` method to handle sync.Once properly (avoid copy-locks)

**Advanced Delegation Logic:**

- `buildFieldsMap()` with ancestor traversal and delegation logic
- Helper methods: `buildFieldsMapFromPath()`, `cacheIfNeeded()`, `populateFromPath()`
- `getFieldsMap()` returns map and freshness indicator
- Use `maps.Copy()` for efficient map operations

### New Ancestor Delegation System (internal/loglet_ancestor.go)

**Complete Delegation Infrastructure (+77 lines):**

- `logletAncestorInfo` struct for delegation state management
- `findCachedAncestor()` to traverse loglet chains and locate cached ancestors
- `tryAncestorCaching()` and `shouldTryAncestorCaching()` for intelligent caching
- `canDelegate()` logic for optimal performance decisions
- `getBaseMap()` with conditional ancestor caching

### Handler Architecture Transformation

**Major Refactoring Pattern (+13/-55 lines total):**

All handlers converted from manual iteration to optimized cached access:

- **cblog** (+1/-15): Replace 15-line iteration with single FieldsMap() call
- **logr** (+2/-10): Cached map with sorted key processing
- **logrus** (+2/-8): Direct logrus.Fields() type conversion
- **mock** (+1/-9): Simplified field recording
- **zap** (+3/-5): Efficient zap.Field slice creation
- **zerolog** (+4/-8): Direct zerolog field addition

**Transformation Example:**

```go
// OLD (15 lines): Manual iteration
if n := l.loglet.FieldsCount(); n > 0 {
    iter := l.loglet.Fields()
    m = make(map[string]any, n)
    for iter.Next() {
        k, v := iter.Field()
        m[k] = v
    }
}

// NEW (1 line): Cached access
Fields: l.loglet.FieldsMap(),
```

### Delegation Logic Flow

1. `FieldsMap()` calls `buildFieldsMap()` for intelligent field access.
2. `findCachedAncestor()` traverses parent chain to locate cached maps.
3. `getBaseMap()` conditionally caches ancestors with fields for delegation.
4. `canDelegate()` determines if direct delegation is possible.
5. Either delegate to cached ancestor or build new map from traversal path.

### Performance Benefits

- **Delegation**: Eliminates redundant map construction for loglets without fields.
- **Memory sharing**: Related loglets share maps when field sets are identical.
- **Proactive caching**: Prevents redundant map construction in ancestor chains.
- **Scalability**: Efficient handling of deep loglet chains.

### New FieldsMapCopy() API

```go
// Returns modifiable copy with optional excess capacity
func (ll *Loglet) FieldsMapCopy(excess int) map[string]any
```

- Provides modifiable field maps without affecting original loglet.
- Supports excess capacity for planned field additions.
- Maintains nil-safety consistent with FieldsMap().

### Comprehensive Test Coverage

- **TestLogletAncestorInfo** (334 lines): Tests delegation logic and state management.
- **TestFindCachedAncestor**: Verifies ancestor traversal in various scenarios.
- **TestAncestorIntegration**: End-to-end delegation workflow testing.
- **TestFieldsMapCopy**: 9 comprehensive sub-tests for new API.
- **TestFieldsMapDelegationCaching**: Verifies delegation caching behaviour.
- Edge case coverage: nil loglets, empty chains, multi-level delegation.

### Performance Benchmarks

New benchmarks in `internal/benchmark_fieldsmap_test.go`:

- **BenchmarkFieldsMapCopy**: Tests FieldsMapCopy performance.
- **BenchmarkFieldsMapCopyWithExcess**: Tests with excess capacity.
- **BenchmarkFieldsMapCopyVsCached**: Compares cached vs uncached sources.
- **BenchmarkFieldsMapCopyVsIteration**: Compares delegation vs manual iteration.
- **BenchmarkFieldsMapDelegationVsIteration**: Tests delegation performance.

### Enhanced Internal Documentation

- **internal/README.md** (+356 lines): Comprehensive documentation of new delegation system.
- Detailed explanation of delegation logic and performance characteristics.
- Examples of usage patterns and optimisation strategies.
- FieldsMapCopy() API documentation with usage examples
- Advanced delegation system explanation with caching heuristics
- Technical documentation for maintainers and contributors.

## Compatibility

- **API**: Maintains full backward compatibility with existing FieldsMap() usage.
- **Behaviour**: FieldsMap() returns same results, but with improved performance.
- **Safety**: Enhanced immutability guarantees and proper nil handling.
- **Handlers**: All handlers benefit from optimisations without code changes.

## Test Plan

- [x] All existing tests pass.
- [x] New delegation system tests comprehensive scenarios.
- [x] FieldsMapCopy API tests cover all edge cases.
- [x] Benchmarks verify performance improvements.
- [x] Handler optimisations maintain functionality.
- [x] Documentation updated with new system details.

## Migration Impact

- **Zero breaking changes**: Existing code continues to work unchanged.
- **Performance improvements**: Automatic for all FieldsMap() usage.
- **New functionality**: FieldsMapCopy() available for modifiable maps.
- **Handler benefits**: All handlers automatically gain performance improvements.

___

### **PR Type**
Enhancement

___

### **Description**
- Implement cached `FieldsMap()` method with delegation optimization

- Add `FieldsMapCopy()` API for modifiable field maps

- Optimize all handlers to use cached field access

- Introduce intelligent ancestor caching system

___

### Diagram Walkthrough

```mermaid
flowchart LR
  A["Loglet.FieldsMap()"] --> B["Check Cache"]
  B --> C["Build & Cache"]
  C --> D["Ancestor Delegation"]
  D --> E["Handler Optimization"]
  E --> F["60x Performance Gain"]
```

<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><details><summary>8 files</summary><table>
<tr>
  <td><strong>cblog.go</strong><dd><code>Replace iterator with cached FieldsMap access</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/darvaza-proxy/slog/pull/128/files#diff-07d30e760b30120b5278f3e252af60bb1ee332a4b190b90e5ec579f224e6d1fa">+1/-15</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>logr.go</strong><dd><code>Optimize field collection using cached map</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/darvaza-proxy/slog/pull/128/files#diff-1a80c7c5faed457a0bcb4581747af2241756d6728706e9808602b685930f407b">+2/-10</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>logrus.go</strong><dd><code>Convert to direct FieldsMap type casting</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/darvaza-proxy/slog/pull/128/files#diff-ba5554efd1a7804862b9622b77de66c65a7e764244b08f43f36c7872e1210cde">+2/-8</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>mock.go</strong><dd><code>Simplify field recording with cached access</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/darvaza-proxy/slog/pull/128/files#diff-625a2b79a7f6e6d4a990140e15406d839ba6717897a12265523511553b370338">+1/-9</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>zap.go</strong><dd><code>Optimize zap.Field slice creation from map</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/darvaza-proxy/slog/pull/128/files#diff-a856627a07f84ea0ff77a00aca21d24b54bc0dbe0ff96c5efbc5e221df67dbb7">+3/-5</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>zerolog.go</strong><dd><code>Replace iteration with direct map access</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/darvaza-proxy/slog/pull/128/files#diff-30d04f4275e7141755ce54efcddc89c1f9588c3fefbde96bb2f195e2a7256f8d">+4/-8</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>loglet.go</strong><dd><code>Implement cached FieldsMap with delegation system</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/darvaza-proxy/slog/pull/128/files#diff-06c05d9619263b50a8414a6215278e7960752a14537339be6d71a2f878d375b2">+150/-11</a></td>

</tr>

<tr>
  <td><strong>loglet_ancestor.go</strong><dd><code>Add ancestor delegation and caching logic</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/darvaza-proxy/slog/pull/128/files#diff-9f3dd959889238c02783283199ee2daa1b2d9c9eac87307a615567d9798f52d0">+77/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Tests</strong></td><td><details><summary>3 files</summary><table>
<tr>
  <td><strong>benchmark_fieldsmap_test.go</strong><dd><code>Add comprehensive FieldsMap performance benchmarks</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/darvaza-proxy/slog/pull/128/files#diff-f9698d6ef5c6e022906b3f20e850f5dd9b49b105cb3823ee83c46b8a9973fd94">+245/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>loglet_ancestor_test.go</strong><dd><code>Test ancestor delegation system functionality</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/darvaza-proxy/slog/pull/128/files#diff-f4dc37c789b6da66daf06bdd4c4fff8ce28c76c9435398411d90ad29a02f0df3">+376/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>loglet_test.go</strong><dd><code>Add FieldsMap and FieldsMapCopy test coverage</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/darvaza-proxy/slog/pull/128/files#diff-9040e7826e981b869c42570bcbf77292e8961d528f3dbdd3b9b1fd611cdfdde2">+564/-11</a></td>

</tr>
</table></details></td></tr><tr><td><strong>Documentation</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td><strong>README.md</strong><dd><code>Document FieldsMap API and delegation system</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/darvaza-proxy/slog/pull/128/files#diff-a1017bdf23ca61e380d4c5737583de77940b04453d9bec88da52834f86633bc8">+356/-0</a>&nbsp; </td>

</tr>
</table></details></td></tr></tr></tbody></table>

</details>

___

## Summary by CodeRabbit

- Refactor
  - Logging handlers now use a unified cached field-map approach for more consistent, deterministic field handling and improved performance (no public API changes).
- Documentation
  - Added internal guidance on logging context, field-access semantics, immutability, delegation, and caching.
- Tests
  - Expanded unit tests covering field-map behavior, delegation, caching, immutability, and integration scenarios.
- Chores
  - Added benchmarks comparing iterator vs map-based field access and copy strategies.

### CODE REVIEW

**Comment by amery** [comment]
> 🧹 Nitpick comments (1)
> > internal/loglet_ancestor_test.go (1)> `135-139`: **Consider verifying map instance identity**

unfortunatelly we can't compare maps with `==` but check now

**Comment by amery** [comment]
# PR Update: Branch Rebased with Thread-Safe Enhancement

## Rebase Summary

Rebased FieldsMap PR from pre-PR #139 base (`10623e3b`) to current main.

## Key Enhancement

**Thread-Safe Caching**: Upgraded from basic FieldsMap caching to proper `sync.Once` implementation for concurrency safety.

This enhancement was only possible after PR #139's `Copy()` method resolved value copying issues that previously prevented adding `sync.Once` fields to Loglet structures.

## Commit Structure

**Before**: 9 commits starting with basic cached FieldsMap() method  
**After**: Same 9-commit structure, but foundational commit now implements thread-safe `sync.Once` caching

The delegation optimization and handler performance improvements remain unchanged - only the underlying caching mechanism was enhanced for proper concurrency support.

File: handlers/cblog/cblog.go [hunk 1]
```diff
@@ -75,24 +75,10 @@ func (l *Logger) Printf(format string, args ...any) {
...
-	}
-
 	l.l.ch <- LogMsg{
 		Message: strings.TrimSpace(msg),
 		Level:   l.loglet.Level(),
-		Fields:  m,
+		Fields:  l.loglet.FieldsMap(),
 		Stack:   l.loglet.CallStack(),
 	}
 }
```
[the remark this answers is not in the document]
Comment (amery) [reply handlers/cblog/cblog.go:84]: you still want to create a new map on every log entry

**Comment by amery** [comment]
# Branch Update Summary

## Test Framework Migration to Core Compliance

**Major refactoring**: Migrated all test functions in `internal/loglet_test.go` and `internal/loglet_ancestor_test.go` to comply with darvaza.org/core testing standards:

- **Replaced manual assertions**: Eliminated 100+ instances of `t.Error`, `t.Errorf`, and `t.Fatal` with proper `core.Assert*` functions
- **Added test helpers**: Added missing `t.Helper()` calls to all 40+ test functions for proper test stack traces  
- **Eliminated reflection**: Replaced reflection-based pointer comparisons (`reflect.ValueOf().Pointer()`) with `core.AssertMustSame` for identity checks
- **Removed unused imports**: Cleaned up `reflect` package imports no longer needed

## Race Condition Elimination

**Critical fix**: Resolved concurrent access issue in FieldsMap delegation algorithm (identified by @coderabbitai):

- **Removed unsafe check**: Eliminated race-prone `if node.fieldsMap != nil` check in `populateFromPath()` that could read unsynchronized memory
- **Fixed delegation logic**: Corrected `canDelegate()` to only allow self-delegation when current loglet has no fields (`len(current.keys) == 0`)
- **Optimized caching**: Updated `cacheIfNeeded()` to use `getFieldsMap()` instead of `FieldsMap()` for better performance
- **Algorithm correction**: Ensured cached ancestors are properly excluded from `pathToRoot` iteration while maintaining correct field population

## Code Quality Improvements

- **Linting compliance**: Fixed line length violation in `buildFieldsMapFromPath()` function signature
- **Code coverage**: Added skipcq comment to suppress false positive in zerolog handler

## Impact

- **Coverage increase**: Internal package coverage improved from 7.9% to 15.1% (+7.2pp)
- **Thread safety**: Eliminated data races while preserving delegation performance optimizations
- **Test reliability**: All tests now follow consistent core testing patterns with proper error handling
- **Maintainability**: Reduced test code complexity by ~190 lines (-368 total, +179 additions)

**Comment by amery** [comment]
Released in darvaza.org/slog v0.8.0: `go get darvaza.org/slog@v0.8.0`
